docs/rfc/rfc-019-config-version.md

```diff
@@ -0,0 +1,269 @@
+that is not possible or practical, we should provide clear, explicit directions
+for what steps need to be taken manually. Moreover, when the node discovers
+incorrect or invalid configuration, we should improve the diagnostics it emits
+so that the operator can quickly and easily find the relevant documentation,
+without having to grep through source code.
+
+## Discussion
+
+By convention, we are supposed to document required changes to the config file
+in the `UPGRADING.md` file for the release that introduces them.  Although we
+have mostly done this, the level of detail in the upgrading instructions is
+often insufficient for an operator to correctly update their file.
+
+The updates vary widely in complexity: Operators may need to add new required
+settings, update obsolete values for existing settings, move or rename existing
+settings within the file, or remove obsolete settings (which are thus invalid).
+Here are a few examples of each of these cases:
+
+- **New required settings:** Tendermint v0.35 added a new top-level `mode`
+  setting that determines whether a node runs as a validator, a full node, or a
+  seed node.  The default value is `"full"`, which means the operator of a
+  validator must manually add `mode = "validator"` (or set the `--mode` flag on
+  the command line) for their node to come up in the correct mode.
+
+- **Updated obsolete values:** Tendermint v0.35 removed support for versions
+  `"v1"` and `"v2"` of the blocksync (formerly "fastsync") protocol, requiring
+  any node using either of those values to update to `"v0"`.
+
+- **Moved/renamed settings:** Version v0.34 moved the top-level `pprof_laddr`
+  setting under the `[rpc]` section.
+
+  Version v0.35 renamed every setting in the file from `snake_case` to
+  `kebab-case`, moved the top-level `fast_sync` setting into the `[blocksync]`
+  section as (itself renamed from `[fastsync]`), and moved all the top-level
+  `priv-validator-*` settings under a new `[priv-validator]` section with their
+  prefix trimmed off.
+
+- **Removed obsolete settings:** Version v0.34 removed the `index_all_keys` and
+  `index_keys` settings from the `[tx_index]` section; version v0.35 removed
+  the `wal-dir` setting from the `[mempool]` section, and version v0.36 removed
+  the `[blocksync]` section entirely.
+
+While many of these changes are mentioned in the config section of the upgrade
+instructions, some are not mentioned at all, or are hidden in other parts of
+the doc. For instance, the v0.34 `pprof_laddr` change was documented only as an
+RPC flag change. (A savvy reader might realize that the flag `--rpc.pprof_laddr`
+implies a corresponding config section, but it omits the related detail that
+there was a top-level setting that's been renamed).  The lesson here is not
+that the docs are bad, but to point out that prose is not the most efficient
+format to convey detailed changes like this.
+
+### Concrete Steps
```

Comment:
I'd like to propose that we ship this as a sub-command within the `tendermint` binary (and also encourage the SDK to embed this subcommand in their own CLI handler so that chain developers don't need to ship an additional binary (or have upgrade instructions that depend on having (a specific version of) go installed or configured.)

Reply:
I'd like to do that, but would prefer not to for the first release, until we have had some production use to validate that it works on real configs rather than just clean initial configs.

Putting it in the TM CLI just makes it harder to experiment and fix bugs—at this stage I'd like to be able to fix things without having to release a new Tendermint. Later, once we have more confidence, I agree we ought to bake it in. 

I _definitely_ don't want the SDK trying to make a programmatic layer out of this yet.

Agreed that it's a little less convenient to set up at first, but there's no specific Go version requirement and it's easily usable with `go get`.